async-iterator.md

```diff
@@ -0,0 +1,904 @@
+- the method yields control back to the background execution (when reaching an `await`).
+Only when the method completes with a `return` is the control fully returned to the caller.
+
+The execution of async-iterator methods mixes both of those patterns, with the control shifting between caller (repeatedly calls `MoveNextAsync()`), method (gives control to the caller when reaching a `yield return` and to the background execution when reaching an `await`) and background execution (resumes the method after an `await`).
+
+
+## Lowering
+
+I'll describe the process of **lowering** async-iterator methods, that is replacing awaits and yields with simpler primitives.
+
+I'll start with many techniques already used for async methods (await suspensions, extracting locals to fields, spilling awaits, dispatch blocks).
+Then I'll explain `yield return` suspensions, disposal and yield breaks.
+
+
+### `await`
+
+`await expr` in async-iterator methods is lowered as:
+```C#
+... code before `await expr` ...
+
+{
+    var awaiterTemp = <expr>.GetAwaiter();
+    if(!awaiterTemp.IsCompleted)
+    {
+        /* set state to N */
+        /* save awaiter temp */
+        /* kick off background execution */
+        return;
+
+        resumeLabelN:
+        /* reset state */
+        /* restore awaiter temp */
+    }
+}
+
+// continue method execution with the result
+... awaiterTemp.GetResult()
+
+... code after `await expr` ...
+```
+
+Note this is the same pattern used to lower awaits in async methods.
+
+As you can see, we're introducing a suspension and resumption in the middle of the method.
+The suspension is the code leading up to the `return` statement. Each suspension is identified by a number `N`.
+The resumption is the code starting from label `resumeLabelN:`.
+
+You might wonder how execution resumes from that label. The lowering also adds logic at the beginning of the method to jumps to `resumeLabelN:` when the **state** is `N`.
+So `{ ... method body... }` is lowered to start with a **dispatching** `switch` statement:
+```C#
+{
+    switch (state)
+    {
+        ...
+        case N:
+            goto resumeLabelN;
+        ...
+    }
+
+    ... rest of lowered method body, including the suspension for state N and the `resumeLabelN` label ...
+}
+```
+
+When either the caller or background execution need to move the method forward, execution will resume from the right label because the state variable was set to `N` and the dispatching logic will jump to that label.
+
+Because there is a cost to starting background execution, suspending and resuming, the lowering pattern is optimized: if the task was already completed, we avoid that overhead and instead just move ahead with the result.
+
+
+#### Locals
+
+Although I won't go into much details on this, it is worth mentioning that all the local variables in the method are converted to fields on a compiler-generated type, so they maintain their values across suspensions/resumptions.
```

Comment:
"all the local variables in the method are converted to fields"... all?  Do we not optimize this to avoid lifting locals that don't need to be?

Reply:
Indeed, only locals that are used across suspensions get hoisted to a field. Thanks